src/Language/PureScript/AST/Declarations.hs

```diff
@@ -325,6 +325,33 @@ instance Eq DeclarationRef where
   (ReExportRef _ mn ref) == (ReExportRef _ mn' ref') = mn == mn' && ref == ref'
   _ == _ = False
 
+instance Ord DeclarationRef where
+  TypeRef _ name dctors `compare` TypeRef _ name' dctors' = compare name name' <> compare dctors dctors'
```

Comment:
Does this "short-circuit" if the `name` comparison is non-EQ and skip the `dctors` comparison?

Would it be more efficient to write it like this:
```hs
TypeRef _ name dctors `compare` TypeRef _ name' dctors'
  | nameCmp == EQ = compare dctors dctors'
  | otherwise = nameCmp
  where nameCmp = compare name name'
```
Not sure if the branching and readability costs outweigh the redundant `dctors` comparison cost. Possibly another job for profiling.

Reply:
> Does this "short-circuit" if the name comparison is non-EQ and skip the dctors comparison?

Yes it does. This relies on Haskell's laziness, so it won't work the same way in PureScript. The `Semigroup Ordering` instance in Haskell is defined like this:

```haskell
instance Semigroup Ordering where
    LT <> _ = LT
    EQ <> y = y
    GT <> _ = GT
```

and the `_` matches won't be evaluated. You can verify this in the repl:
```
λ: LT <> error "this is fine"
LT
λ: EQ <> error "this is not fine"
*** Exception: this is not fine
CallStack (from HasCallStack):
  error, called at <interactive>:2:7 in interactive:Ghci2
λ: GT <> error "this is fine"
GT
```